js/app.js

```diff
@@ -32,57 +30,136 @@ function rPG() { // the notorious Random Product Generator (this isn't nearly as
     // generate disaplyQty number of randomly selected unique product images to be displayed to user
     while (products.length < displayQty) {
         // returns a whole number 0 - index of last item in allProducts array
-        let index = Math.floor(Math.random() * prodObjs.length);
-        while (!products.includes(prodObjs[index])) {
-            products.push(prodObjs[index]);
-            prodObjs[index].countShown++;
+        let index = Math.floor(Math.random() * allProducts.length);
+        // prevent repeating a product that's 1) already in this voting round or 2) displayed in the previous voting round
+        while (!products.includes(allProducts[index]) && !lastDisplayed.includes(allProducts[index])) {
+            // add randomly selected product to this voting round
+            products.push(allProducts[index]);
+            // increment display counter for randomly selected product
+            allProducts[index].countShown++;
         }
     }
+
     // display each of the selected products and insert attribute values
     for (let i = 0; i < products.length; i++) {
         let img = document.getElementById(`productImage${i}`);
         img.src = products[i].filePath;
         img.alt = products[i].name;
         img.title = products[i].name;
     }
+    lastDisplayed = products;
 }
 
 // function to display View Results button after voteCount === maxVotes, and add event listener for said button
 function displayButton() {
     // display button viewResultsBtn
     document.getElementById('viewResults').innerHTML = '<input type="button" id="viewResultsBtn" value="View Results"/>';
-    // event listener to invole renderResults when button is clicked
+    // event listener to invoke renderResults when button is clicked
     let viewResultsBtn = document.getElementById('viewResultsBtn');
     viewResultsBtn.addEventListener('click', renderResults);
 }
 
+// function to render chart of voting results
+function renderChart() {
+    let productNames = [];
+    for (let i = 0; i < allProducts.length; i++) {
+        productNames.push(allProducts[i].name);
+    }
+    let voted  = [];
+    for (let i = 0; i < allProducts.length; i++) {
+        voted.push(allProducts[i].countVoted);
+    }
+    let shown  = [];
+    for (let i = 0; i < allProducts.length; i++) {
+        shown.push(allProducts[i].countShown);
+    }
```

Comment:
Since you're iterating through the same array each time, consider combining all of this into one loop, so you iterate three times less.

Reply:
It's seems so obvious now that you say it. Thanks for the tip.